### ISSUE TITLE
Reproduction of Lesion Segmentation Model

### ISSUE THREAD

**Issue by CloseChoice** [CloseChoice/neuroimaging-go-brrrr#19]
Would be pretty cool if we can reproduce these results:
 - https://arxiv.org/html/2503.05531v1

A first step would be to figure out the data sets that include lesion, then combine them into one training dataset. Afterwards checking out MONAI (https://github.com/Project-MONAI/MONAI) to use their structure for a potential model candidate.

**Comment by The-Obstacle-Is-The-Way** [comment]
Hey @CloseChoice - I went ahead and started implementing the paper you linked here.

**Repo:** https://github.com/The-Obstacle-Is-The-Way/arc-meshchop

It's the Fedorov et al. MeshNet approach - targeting 0.876 DICE with 147K params, training on ARC lesion data from HuggingFace Hub.

Happy to add you or anyone else as contributors!

I put it in a separate repo since I felt this one is better suited as a consumption/production pipeline for datasets. Adding training code would cause a dependency explosion (PyTorch, etc.). 

If you feel differently, let me know - we could always try merging it here. Just didn't seem like best practice to me.

**Comment by CloseChoice** [comment]
No, that sounds good. Just to carry over the conversation from Discord:

Thoughts from reading: https://arxiv.org/html/2503.05531v1

- what is "mri_convert"? This is the paper: https://surfer.nmr.mgh.harvard.edu/ftp/articles/1999/1999_-_Dale_et_al._-_NeuroImage.pdf but haven't seen any github link there. Though it might be available in the freesurfer repos
- From the paper: "We utilized Sampling Perfection with Application optimized Contrast using different flip angle Evolution (SPACE) sequences with x2 in plane acceleration (115 scans) and without acceleration (109 scans), while excluding the turbo-spin echo T2-weighted sequences (5 scans) to maintain homogeneity in imaging protocols." <- tbh I have no idea what this means, AI might be able to help out here but we somehow need to understand that and make sure we are doing the right thing
- we would need to implement the dilation 3d convolution, which should be relatively easy I assume
- why do they need a 80GB GPU? These are small models and they mention they didn't store the checkpoints on the GPU due to size (or at least that's the way I read it)? I mean it makes sense to not store model checkpoints on the GPU but for such a large GPU size shouldn't be the limiting factor

Maybe starting with: https://github.com/Project-MONAI/tutorials/tree/main/2d_segmentation/torch is a first good step here

**Comment by The-Obstacle-Is-The-Way** [comment]
Hey @CloseChoice, great questions - these are the things I dug into while implementing:

---

### 1. What is `mri_convert`?

It's a command-line tool from **FreeSurfer** (the neuroimaging suite), not a standalone repo. You can find it here: https://surfer.nmr.mgh.harvard.edu/

The paper uses it for preprocessing: resample to 256×256×256 @ 1mm isotropic + normalize intensities to 0-1.

In arc-meshchop, I implemented a **pure Python equivalent** using nibabel + scipy that produces the same output - so no FreeSurfer dependency needed. See: [`src/arc_meshchop/data/preprocessing.py`](https://github.com/The-Obstacle-Is-The-Way/arc-meshchop/blob/main/src/arc_meshchop/data/preprocessing.py)

---

### 2. The SPACE/TSE sequences paragraph

**SPACE** and **TSE** are just different MRI acquisition protocols - both produce T2-weighted images that look similar, but with slightly different signal characteristics.

The ARC dataset has 228 scans total:
- 115 acquired with SPACE (2x acceleration)
- 108 acquired with SPACE (no acceleration)  
- 5 acquired with TSE (Turbo Spin Echo)

The paper authors excluded the 5 TSE scans to keep the data homogeneous - they wanted all training data from the same acquisition protocol. So they trained on **224 SPACE-only samples**.

The acquisition type is encoded in the BIDS filenames on OpenNeuro:
- `acq-spc3p2` → SPACE 2x
- `acq-spc3` → SPACE no acceleration
- `acq-tse3` → TSE (excluded)

I documented all of this in detail here: [`docs/KNOWN_ISSUES.md`](https://github.com/The-Obstacle-Is-The-Way/arc-meshchop/blob/main/docs/KNOWN_ISSUES.md)

---

### 3. Dilated 3D convolution

Already implemented. The paper's key innovation is a **symmetric 10-layer dilation pattern**:

```
1 → 2 → 4 → 8 → 16 → 16 → 8 → 4 → 2 → 1
     (encoder)          (decoder)
```

This mimics U-Net's encoder-decoder structure without any downsampling, upsampling, or skip connections. Just dilated 3D convs with careful padding.

See: [`src/arc_meshchop/models/meshnet.py`](https://github.com/The-Obstacle-Is-The-Way/arc-meshchop/blob/main/src/arc_meshchop/models/meshnet.py)

---

### 4. Why 80GB GPU?

The 80GB A100 was for the **baseline models** they compared against, not MeshNet:

| Model | Parameters |
|-------|------------|
| MeshNet-26 | 147K ← tiny! |
| U-MAMBA | 7.3M |
| MedNeXt-M | 17.5M |
| V-Net | 45.6M |
| UNETR | 95.8M |

The paper says: *"We used layer checkpointing for models that could not fit into GPU memory."* - that's for the big baselines.

**MeshNet is so small it runs on Apple Silicon (MPS).** I've been developing on a MacBook. The whole point of the paper is that you *don't* need massive hardware for MeshNet.

---

### 5. Re: MONAI

I'd actually recommend **not** using MONAI if the goal is paper parity. The paper uses vanilla PyTorch + Hydra - no MONAI mentioned anywhere. Introducing MONAI could cause subtle differences in data loading, augmentation, or training behavior that make it harder to match their exact results.

For reproducibility, I'm sticking to their exact stack. Once we hit paper parity (0.876 DICE), then we could compare to a MONAI implementation if we want. 

---

The repo is pretty much ready for training - just waiting on a small upstream fix to the HuggingFace dataset metadata so we can filter to the exact 224 SPACE-only samples. Should be done soon.

Let me know if you want to dig into any of this further.

**Comment by CloseChoice** [comment]
great, thanks a lot for the response, that makes sense.

I mean we could also simply install the datasets branch you are waiting to be merged to start training. I might give it a shot sometime in the upcoming week. Which branch is it?

**Comment by The-Obstacle-Is-The-Way** [comment]
Yep no problem: 

The main blocker isn’t really a merge right now. The initial ARC dataset upload on Hugging Face is missing a few metadata fields that the paper’s processing expects.
Those fields do exist in the original OpenNeuro release, so if you want to experiment locally you can work directly from OpenNeuro today.

In my fork I’ve been updating the production pipeline to reliably generate those metadata fields and re-upload the ARC dataset with the complete metadata to the Hub, since the whole goal of this pipeline is to consume datasets from Hugging Face end to end. I’m in the last stretch of those fixes now.

Once the updated dataset is up, training should be straightforward. If you want to test sooner, I've been working on my fork here: https://github.com/The-Obstacle-Is-The-Way/neuroimaging-go-brrrr

**Comment by The-Obstacle-Is-The-Way** [comment]
> I mean we could also simply install the datasets branch you are waiting to be merged to start training. I might give it a shot sometime in the upcoming week. Which branch is it?

Quick clarification since "datasets" can mean a few things here:

Originally we were pushing PRs to HuggingFace `datasets` itself (Nifti, BIDS support), but they've been slow to merge and it's unclear if they want all of it upstream. So I've been building out `neuroimaging-go-brrrr` as the canonical consumption/production extension for neuroimaging, rather than waiting indefinitely on those PRs.

I documented the architecture here to help clarify: [ARCHITECTURE.md](https://github.com/The-Obstacle-Is-The-Way/neuroimaging-go-brrrr/blob/main/ARCHITECTURE.md)

Since I wasn't sure if anyone really wanted to contribute anymore, I've just been updating my local fork, which has now significantly diverged from this repo. 

If you want, I'm happy to just merge it directly with all the bug fixes. Or I can keep it separate if that's cleaner. Either works for me.
